README
---------------
| DESCRIPTION |
---------------

Title    : eLuna Graph System
Version  : 1.10
Released : 2017-06-23
Website  : https://github.com/stephanedupont/eLunaGraphSystem
Demo     : http://graphs.eluna.org
Author   : Stephane Dupont
License  : GNU General Public License (see LICENSE)

eLuna Graph System is an application written in Perl based on
RRDTool. Its aim is collection and then presentation in graphic
form of data to aid in the monitoring of a machine. By default,
the application allows monitoring of the load, CPU usage, memory
use, number of processes, amount of data transmitted via eth0
(In/Out) and disk space used in /.


----------------
| PREREQUISITE |
----------------

- root access on the machine to monitor
- An http server (preferably Apache)
- RRDTool (preferably version >= 1.2)
- Perl (preferably version >= 5.8)
  with the following modules installed:
  * DateTime
  * HTML::Template::Expr


----------------
| INSTALLATION |
----------------

- Download the latest version here:

  http://graphs.eluna.org/releases/latest/eluna_graph_system.tar.gz

- Decompress the archive into a folder accessible via Apache.

  Example:
    mv eluna_graph_system.tar.gz /var/www/graphs/
    cd /var/www/graphs/
    tar -xzf eluna_graph_system.tar.gz

- Perform if necessary a chown and/or a chmod on all scripts so
  that they are readable and executable by the user/group used by
  the http server. The 'graphs' folder must also be writeable by the
  http server. In the majority of cases, you have nothing to do.

- Schedule the execution of the 'update.pl' script every 5 minutes
  with crontab.

  Example:
    */5 * * * * root /var/www/graphs/update.pl

- Configure Apache so that the 'index.pl' script is interpreted correctly
  as a perl script and that is viewed as 'DirectoryIndex'.

  Example:
    <Directory /var/www/graphs/>
      AddHandler cgi-script .pl
      Options +ExecCGI
      DirectoryIndex index.pl
    </Directory>

- Ensure that the '.htaccess' file containing 'deny from all' and contained
  in the 'rrd' folder is taken into account by Apache. The 'rrd' folder must
  not be accessible via http. In the case of an http server other than Apache,
  create a similar protective mechanism.

The application should now work fine. Wait about ten minutes so that the cron
task can be executed at least twice, then try to view the script in a web browser.


-----------------
| CUSTOMIZATION |
-----------------

1/ Configuration
----------------

The config file 'config.pm' contains a number of options like the name of
the machine, the format of the pictures generated, the size of the images
generated, the colours of the images generated, the default view, etc.


2/ Application appearance/texts
-------------------------------

The application uses the Perl HTML::Template::Expr module, itself being an
extension of the Perl HTML::Template module, which allows for definition of
an external HTML template to manage the graphical appearance, the layout
and the texts. This template is located in the 'template' folder.

For more information please consult the documentation for these two modules
here:

  http://search.cpan.org/~samtregar/HTML-Template/lib/HTML/Template.pm
  http://search.cpan.org/~samtregar/HTML-Template-Expr/Expr.pm


3/ Add/customize elements
-------------------------

Notes:

- It is strongly advised when adding elements to base these on an existing
  element rather than starting from scratch.

- It is strongly advised that you read the documentation on RDDTool before
  customizing or adding elements. This documentation, as well as tutorials,
  are available on the RRDTool site:

    http://oss.oetiker.ch/rrdtool/doc/index.en.html


How it works
------------

 Each element is composed of a folder placed in the 'rrd' directory. This
 folder must contain a create file ('create.sh'), an update file ('update.sh'
 or 'update.pl') and a file for generating the graphs ('graph.pm').


Folder name
-----------

 The name of the folder must be the element ID, possibly preceded by an
 integer and by the character '_'. For example, for an element which one
 wishes to identify by 'cpu', the name of the folder may be 'cpu' or, for
 example, '123_cpu'.

 When the element ID is preceded by an integer this affects/determines
 the display order as the elements are displayed in alphabetical order of
 the folders containing them (and not in the alphabetical order of their IDs).


'create.sh'
-----------

 The file 'create.sh' is responsible for the creation of the Round Robin
 Database. It is a shell file containing an RRD create command. The RRD
 created must be named 'id.rrd', where 'id' is the element ID.

 For various reasons, it is desirable to specify the same RRA (Round Robin
 Archives) and the same step value for each of the application elements.

 Example:
   Creation of a Round Robin Database corresponding to the 'cpu' element
   and intended for the storage of four DS (Data Sources): 'user', 'system',
   'nice' and 'idle'.

   #!/bin/bash

   rrdtool create cpu.rrd \
     --start `date +%s` \
     --step 300 \
     DS:user:COUNTER:600:0:U \
     DS:system:COUNTER:600:0:U \
     DS:nice:COUNTER:600:0:U \
     DS:idle:COUNTER:600:0:U \
     RRA:AVERAGE:0.5:1:2016 \
     RRA:AVERAGE:0.5:6:1344 \
     RRA:AVERAGE:0.5:24:732 \
     RRA:AVERAGE:0.5:144:1460


'update.sh' / 'update.pl'
-------------------------

  The file 'update.sh' is responsible for updating the Round Robin Database
  associated with the element. The execution of this script must therefore
  insert into the RRD a new value for all the DS (Data Sources) that are
  defined in the create file.

  Note that it is possible to replace this 'update.sh' shell file with a
  perl file 'update.pl'. If the first is not found, the second will be
  executed.

  Example:
   Updating of the RRD associated with the 'cpu' element by a perl script

   #!/usr/bin/perl

   $dummy = `cat /proc/stat | grep "^cpu "`;
   $dummy =~ /(.*) (.*) (.*) (.*) (.*)/;

   system("rrdtool update cpu.rrd -t user:system:nice:idle N:$2:$3:$4:$5");


'graph.pm'
----------

  The file 'graph.pm' contains parameters for the display of the graph
  associated with the element. This must include:

  - The graph title, contained in the variable $GRAPH_TITLES{'id'} (where
    'id' is the element ID). If {#server#} is present in this title it will
    be substituted by the server name.

  - A group of parameters that will be transferred to RDDTool. This group
    of parameters must be contained in the variable $GRAPH_CMDS{'id'} (where
    'id' is the element ID).

    Some substitutions will be effected among these parameters. See the
    following list:

    {#server#}    : Server name
    {#path#}      : Element folder path
    {#color1#}    : Draw color 1
    {#color2#}    : Draw color 2
    {#color3#}    : Draw color 3
    {#color4#}    : Draw color 4
    {#color5#}    : Draw color 5
    {#dcolor1#}   : Gradient color 1
    {#dcolor2#}   : Gradient color 2
    {#dcolor3#}   : Gradient color 3
    {#linecolor#} : Lines color

    Note that certain parameters are managed by the application and do not
    require to be specified here. These parameters are:

    - the name of the image file generated
    - the type of the image file generated
    - the start date of the graph (--start)
    - the end date of the graph (--end)
    - the graph dimensions (--w and --h)
    - graph comment

    Example:
    'graph.pm' file associated with the 'cpu' element of previous examples

    $GRAPH_TITLES{'cpu'} = "{#server#} - CPU Usage";
    $GRAPH_CMDS{'cpu'} = <<"CPU_GRAPH_CMD";
    --title "{#server#} - CPU Usage"
    --vertical-label="Percent"
    --lower-limit 0 --upper-limit 100
    DEF:user={#path#}cpu.rrd:user:AVERAGE
    DEF:system={#path#}cpu.rrd:system:AVERAGE
    DEF:nice={#path#}cpu.rrd:nice:AVERAGE
    DEF:idle={#path#}cpu.rrd:idle:AVERAGE
    CDEF:total=user,system,+,nice,+,idle,+
    CDEF:p_user=user,total,/,100,*
    CDEF:p_system=system,total,/,100,*
    CDEF:p_nice=nice,total,/,100,*
    CDEF:mysum=p_user,p_system,+,p_nice,+
    AREA:p_user{#dcolor1#}:"User   "
    GPRINT:p_user:LAST:"Current\\: %5.2lf%%  "
    GPRINT:p_user:AVERAGE:"Average\\: %5.2lf%%  "
    GPRINT:p_user:MAX:"Max\\: %5.2lf%%\\n"
    STACK:p_system{#dcolor2#}:"System "
    GPRINT:p_system:LAST:"Current\\: %5.2lf%%  "
    GPRINT:p_system:AVERAGE:"Average\\: %5.2lf%%  "
    GPRINT:p_system:MAX:"Max\\: %5.2lf%%\\n"
    STACK:p_nice{#dcolor3#}:"Nice   "
    GPRINT:p_nice:LAST:"Current\\: %5.2lf%%  "
    GPRINT:p_nice:AVERAGE:"Average\\: %5.2lf%%  "
    GPRINT:p_nice:MAX:"Max\\: %5.2lf%%\\n"
    LINE1:mysum{#linecolor#}
    CPU_GRAPH_CMD

    1; # Return true

CHANGELOG
| ----- |
| v1.09 |
| ----- |

- You can now download the raw RRDtool data used to generate the graphs in a CSV file
- Links added to download the graphs as PNG files
- Images filenames are now UIDs


| ----- |
| v1.08 |
| ----- |

- You can now choose not to display some of the DS in the graphs
- Option to add UID to img urls (to be sure IE refreshes images and doesn't use the cache)
- Configuration is now in a separated file
- UPGRADE file added
- Use of JQuery
- A few minor fixes
- Code cleaning


---------
| v1.07 |
---------

- Datasource type is now DERIVE for cpu et eth0
- Label of eth0 graphs updated
- Order of cpu graph parameters changed
- Updated the RRDTool link to the correct URL (http://oss.oetiker.ch/rrdtool/)

To update from a previous version, run these commands in eLuna Graph System root folder:

rrdtool tune rrd/05_eth0_out/eth0_out.rrd -d out:DERIVE:600:0:U
rrdtool tune rrd/06_eth0_in/eth0_in.rrd -d in:DERIVE:600:0:U
rrdtool tune rrd/02_cpu/cpu.rrd -d user:DERIVE:600:0:U
rrdtool tune rrd/02_cpu/cpu.rrd -d nice:DERIVE:600:0:U
rrdtool tune rrd/02_cpu/cpu.rrd -d system:DERIVE:600:0:U
rrdtool tune rrd/02_cpu/cpu.rrd -d idle:DERIVE:600:0:U

Thanks to Ernst Lobsiger for these corrections/improvements.


---------
| v1.06 |
---------

- Bug corrected in rrd/02_cpu/update.pl, thanks to Ernst Lobsiger
  ("system" and "nice" values were exchanged)


---------
| v1.05 |
---------

- New view added : "Last Two Days"
- View labels updated
- robots.txt file added
- "Custom View" graphs are now displayed as "Zoomed" graphs
- The "Custom View" form will now also appears in "Zoom" mode
- The page will no longer refresh itself if in "Custom View" mode
- New configuration variable $DELAY_BETWEEN_CUSTOM


---------
| v1.04 |
---------

- Timezone problem corrected


---------
| v1.03 |
---------

- Bug fixes in rrd/03_mem and in rrd/04_process


---------
| v1.02 |
---------

- English documentation added
- 'ifconfig' replaced by '/sbin/ifconfig' (thanks to Ben West)
- Add a test to prevent divide by 0 error when used in a no swap
  memory system (thanks to Ben West)
- Timezone on Debian is no longer a problem (thanks to Marco Barcella)


---------
| v1.01 |
---------

- RRDTool compatibility bug correction
  (use of -w/-h in place of --w/--h)


---------
| v1.00 |
---------

First release.


index.pl
#!/usr/bin/perl

 #                                                        #
 # ------------------------------------------------------ #
 # eLuna Graph System                                     #
 # ------------------------------------------------------ #
 # File     : index.pl                                    #
 # Author   : Stephane Dupont                             #
 # Version  : 1.10                                        #
 # Released : 2017-06-23                                  #
 # Summary  :                                             #
 #   Script that display graphs                           #
 #   To be used in a web navigator                        #
 # ------------------------------------------------------ #
 #                                                        #


# --------------------------------------------------------------- #
# Note : there's two types of ID's for RRDs,                      #
# - ID : for example load, cpu, ...                               #
# - SID (Sort ID) : for example 01_load, 02_cpu, ...              #
# ID can be equal to SID if no integer is added in the foldername #
# --------------------------------------------------------------- #


die ("This script must be executed as a web script !") if !$ENV{'HTTP_HOST'};

use POSIX qw(strftime);
use CGI;
use CGI::Cookie;
use DateTime;
use Time::HiRes qw(gettimeofday);
use HTML::Template::Expr;


# ---------------- #
# Load config file #
# ---------------- #

require 'config.pm';


# ----------------- #
# Some variables... #
# ----------------- #

my $cgi = new CGI;
our (%GRAPH_CMDS, %GRAPH_TITLES);
my ($title, $back_link, $summary, $display_custom_view, $cookie,
    $custom_view, $ctime1, $ctime2, @elgraphs, @rrds, @trace);


# -------------------------- #
# Print something at the end #
# of page source code        #
# -------------------------- #
# Param  : info              #
# Return : none              #
# -------------------------- #

sub elTrace
{
  push @trace, "<!-- Trace: $_[0] -->\n";
}


# ------------ #
# Return a UID #
# ------------ #

sub getUID
{
  ($s, $usec) = gettimeofday();
  $s . $usec . $$ . int rand(1000) . int rand(1000);
}


# ------------------------------------------------------- #
# Convert a datetime from 'yyyy-mm-dd hh:mm' to epochtime #
# ------------------------------------------------------- #
# Param  : datetime_string                                #
# Return : epochtime                                      #
# ------------------------------------------------------- #

sub getEpoch
{
  if ($_[0] =~ /([0-9]{4})-([0-9]{2})-([0-9]{2}) ([0-9]{2}):([0-9]{2})/) {
    my $dt = DateTime->new(
      year      => $1,
      month     => $2,
      day       => $3,
      hour      => $4,
      minute    => $5,
      time_zone => "local"
    );
    $dt->epoch;
  } else {
    0;
  }
}


# ---------------------------------- #
# Return RRD List (array of strings) #
# Strings are as SID                 #
# ---------------------------------- #
# Param  : none                      #
# Return : array_of_rrd_strings      #
# ---------------------------------- #

sub getRRDList
{
  opendir(DIR, $RRD_DIR);
  my @rrds =
    grep { -d $RRD_DIR.$_ }
      grep { ( $_ ne '.' ) and ( $_ ne '..' ) }
          readdir DIR;
  closedir DIR;
  sort @rrds;
}


# ------------------------------------ #
# Return RRD List (sid, name, current) #
# Used in the template file to display #
# menu when not in summary view        #
# To be easier to use in it, 'sid' key #
# will be named 'id'                   #
# ------------------------------------ #
# Param  : current_rrd_sid             #
# Return : ref_on_array_of_hach_refs   #
# ------------------------------------ #

sub getRRDItems
{
  my @result;
  foreach my $rrd (@rrds) {
    my %th;
    $th{'id'} = $rrd;
    require $RRD_DIR.$rrd.'/graph.pm';
    my $tt = $rrd; $tt =~ s/^[0-9]+_//; 
    $th{'name'} = $GRAPH_TITLES{$tt};
    $th{'name'} =~ s/\{#server#\}[ -]*//g;
    $th{'current'} = 1 if $th{'id'} eq $_[0];
    push @result, \%th;
  }
  \@result;
}


# -------------------------------------- #
# Return the 'cds' (Custom Data Sources) #
# string for a RRD or empty string if    #
# not relevant (1..1, 0..0)              #
# Example : 110, 101, 01, ...            #
# -------------------------------------- #
# Param  : rrd_sid                       #
# Return : string                        #
# -------------------------------------- #

sub getCDSString
{
  my $str;
  if ($cookie && $cookie->name eq $_[0]) {
    $str = $cookie->value;
  } else {  
    $str = $cgi->cookie($_[0]);
  }
  if (($str =~ /^[0-1]+$/) && ($str =~ /0/) && ($str =~ /1/)) {
    return $str;
  } else {
    return '';
  }
}


# -------------------------------------- #
# Return DS List (id, name, type, state) #
# for a given RRD SID                    #
# -------------------------------------- #
# Param  : rrd_sid                       #
# Return : ref_on_array_of_hach_refs     #
# -------------------------------------- #

sub getDSItems
{
  my @result;

  my $rrd_id  = $_[0];
  $rrd_id =~ s/^[0-9]+_//;

  my $cmd = $GRAPH_CMDS{$rrd_id};

  my $cds = getCDSString($_[0]);

  my @lines = split(/\n/, $cmd);
  my $counter = 0;
  my $ctype = '';
  foreach my $line (@lines) {
    if ($line =~ /^\s*(AREA|STACK|LINE)\:([^\:]*)\:\"(.*)\"/) {
      $counter++;
      my $id = $2;
      my $name = $3;
      my $type = $1;
      my $stack = 0;
      if ($type eq 'STACK') {
        $type = $ctype;
        $stack = 1;
      } else {
        if ($line =~ /\:STACK/) { # New syntax for STACK, example:
          $stack = 1;             # AREA:value[#color][:[legend][:STACK]]
        }
        $ctype = $type;
      }
      $id =~ s/\{.*\}//g;
      $id =~ s/^\s+|\s+$//g;
      $name =~ s/^\s+|\s+$//g;
      $type =~ s/^\s+|\s+$//g;
      $state = 1;
      $state = 0 if $cds =~ /^0/;
      $cds =~ s/^.//;
      push @result, { 'id' => $id, 'name' => $name , 'type' => $type, 'stack' => $stack, 'state' => $state};
    }
  }
  shift(@result) if $#result==0; # we don't need DS list if only 1 item
  \@result;
}


# ----------------------------------------- #
# Customize a graph command to display only #
# some of the DS                            #
# ----------------------------------------- #
# Param  : list of DS (see getDSItems)      #
# Param  : command to customize (string)    #
# Return : customized command (string)      #
# ----------------------------------------- #

sub customizeRRDCmd
{
  my @dss = @{$_[0]};
  my $cmd = $_[1]; 
  my $result = '';
  my %ids;

  my $ctype = '';
  foreach my $ds (@dss) {
    if ($$ds{'state'} == 1) {
      $ids{$$ds{'id'}} = 1;
      if (($$ds{'type'} ne $ctype) && $$ds{'stack'}) {
        $ids{$$ds{'id'}} = $$ds{'type'};
      }
      $ctype = $$ds{'type'};
    }
  } 

  my @lines = split(/\n/, $cmd);
  foreach my $line (@lines) {
    if ($line =~ /^\s*(AREA|STACK|LINE|PRINT|GPRINT)\:([^\:]*)\:/) {
      $id = $2;
      $id =~ s/\{.*\}//g;
      $id =~ s/^\s+|\s+$//g;
      if ($ids{$id}) {
         if ($ids{$id} != 1) {
           my $correct = $ids{$id};
           $line =~ s/\:STACK//;
           $line =~ s/^\s*STACK/$correct/;
         }
         $result .= $line."\n";
      }
    } else {
      $result .= $line."\n";
    }
  }

  # Remove non-displayed DS from CDEF sums (used to draw lines on top of areas)
  @lines = split(/\n/, $result);
  $result = '';
  foreach my $line (@lines) {
    if ($line =~ /^\s*(C|V)DEF\:[^\,]+\,([^\,]+\,\+\,)+/) {
      &elTrace('Pb: '.$line);
      foreach my $ds (@dss) {
        if ($$ds{'state'} != 1) {
          my $id = $$ds{'id'};
          $line =~ s/([=\,])$id(\,\+)?\,?/$1/g;  # we remove the element and following '+' sign
        }
      }
      $line =~ s/(=[^\,\:]+)\,\+/$1/; # if $line is '=a,+[...]', we remove ',+'
      &elTrace('Pb: '.$line); 
    }
    $result .= $line."\n";
  }
 
  $result; 
}


# ------------------------------------------------- #
# Create a graph                                    #
# ------------------------------------------------- #
# Param  : rrd_sid                                  #
# Param  : type ('h', 'd', 'b', 'm', 'w', 'y', 'c') #
# Param  : [Optional] time1 (yyyy-mm-dd hh:mm)      # 
# Param  : [Optional] time2 (yyyy-mm-dd hh:mm)      #
# Return : none                                     #
# ------------------------------------------------- #

sub createRRDGraph
{
  my %elh;

  $elh{'rrd_sid'} = $_[0];

  $elh{'rrd_id'} = $_[0];
  $elh{'rrd_id'} =~ s/^[0-9]+_//;

  require $RRD_DIR.$elh{'rrd_sid'}.'/graph.pm';

  $elh{'gtitle'} = $GRAPH_TITLES{$elh{'rrd_id'}};
  $elh{'gtitle'} =~ s/\{#server#\}/$SERVER/g;
  $elh{'gtype'}  = $_[1];

  if ($_[1] ne 'c') {
    my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst);
    do {
      $elh{'time2'} = time();
      ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime($elh{'time2'});
    } until $sec != 0;
    $elh{'time2'} -= ($sec-1);
    for ($i = 0; $i < ($min%5); $i++) {$elh{'time2'} -= 60;}
    $elh{'time1'} = $elh{'time2'} - 14400    if $_[1] eq 'h';
    $elh{'time1'} = $elh{'time2'} - 86400    if $_[1] eq 'd';
    $elh{'time1'} = $elh{'time2'} - 172800   if $_[1] eq 'b';
    $elh{'time1'} = $elh{'time2'} - 604800   if $_[1] eq 'w';
    $elh{'time1'} = $elh{'time2'} - 2592000  if $_[1] eq 'm';
    $elh{'time1'} = $elh{'time2'} - 31536000 if $_[1] eq 'y';
  } else {
    $elh{'time1'} = &getEpoch($_[2]);
    $elh{'time2'} = &getEpoch($_[3]);
  }

  $elh{'uid'} = getUID();

  $elh{'img_src'} = $GRA_DIR.$elh{'uid'}.'.'.lc($IMG_FORMAT);
  
  $elh{'strtime1'} = strftime "%Y-%m-%d %H\\:%M\\:%S", localtime($elh{'time1'});
  $elh{'strtime2'} = strftime "%Y-%m-%d %H\\:%M\\:%S", localtime($elh{'time2'});

  $elh{'dss'} = &getDSItems($elh{'rrd_sid'});

  $elh{'cmd'} = "rrdtool graph ".$elh{'img_src'}." ";
  $elh{'cmd'} .= "--start ".$elh{'time1'}." --end ".$elh{'time2'}." ";
  $elh{'cmd'} .= "--imgformat ".$IMG_FORMAT." --lazy ";
  $elh{'cmd'} .= "-w ".$GRAPH_WIDTH." -h ".$GRAPH_HEIGHT." " if !$cgi->param('type');
  $elh{'cmd'} .= "-w ".$ZOOM_WIDTH." -h ".$ZOOM_HEIGHT." " if $cgi->param('type');
  $elh{'cmd'} .= 'COMMENT:"From '.$elh{'strtime1'}.' To '.$elh{'strtime2'}.'\c" ';
  $elh{'cmd'} .= 'COMMENT:"\n" ';
 
  if (getCDSString($elh{'rrd_sid'}) =~ /0/) {
    $elh{'cmd'} .= customizeRRDCmd($elh{'dss'}, $GRAPH_CMDS{$elh{'rrd_id'}});
  } else {
    $elh{'cmd'} .= $GRAPH_CMDS{$elh{'rrd_id'}};
  }

  $elh{'cmd'} =~ s/[\n\r]+/ /g;
  $elh{'cmd'} =~ s/\{#server#\}/$SERVER/g;
  $elh{'cmd'} =~ s/\{#path#\}/$RRD_DIR$elh{'rrd_sid'}\//g;
  $elh{'cmd'} =~ s/\{#linecolor#\}/$GRAPH_LINECOLOR/g;
  $elh{'cmd'} =~ s/\{#color1#\}/$GRAPH_COLOR1/g;
  $elh{'cmd'} =~ s/\{#color2#\}/$GRAPH_COLOR2/g;
  $elh{'cmd'} =~ s/\{#color3#\}/$GRAPH_COLOR3/g;
  $elh{'cmd'} =~ s/\{#color4#\}/$GRAPH_COLOR4/g;
  $elh{'cmd'} =~ s/\{#color5#\}/$GRAPH_COLOR5/g;
  $elh{'cmd'} =~ s/\{#dcolor1#\}/$GRAPH_DCOLOR1/g;
  $elh{'cmd'} =~ s/\{#dcolor2#\}/$GRAPH_DCOLOR2/g;
  $elh{'cmd'} =~ s/\{#dcolor3#\}/$GRAPH_DCOLOR3/g;

  open my $oldout, ">&STDOUT";
  open STDOUT, ">/dev/null";
  system($elh{'cmd'});
  close STDOUT;
  open STDOUT, ">&", $oldout;

  if (!$cgi->param('rrd')) {
    $elh{'link_type'} = "more";
    $elh{'link_url'} = "index.pl?rrd=".$elh{'rrd_sid'};
  } elsif (!$cgi->param('type')) {
    $elh{'link_type'} = "zoom";
    $elh{'link_url'} = "index.pl?rrd=".$elh{'rrd_sid'}."&type=$_[1]&time1=$_[2]&time2=$_[3]";
  }

  $elh{'display_command'} = $cgi->param('type');

  push @elgraphs, \%elh;
}


# ------------ #
# Get RRD list #
# ------------ #
@rrds = &getRRDList();


# ----------------------------------------------------------- #
# Check validity of cgi parameters to limit security problems #
# ----------------------------------------------------------- #
my $trrd;
if (($cgi->param('type') &&
     !($cgi->param('type') =~ /^[hdbwmyc]$/))
    ||
    ($cgi->param('rrd') &&
     ($trrd = $cgi->param('rrd')) &&
     !(grep /^$trrd$/, @rrds))
    ||
    ($cgi->param('type') && !$cgi->param('rrd'))
    ||
    ($cgi->param('type') eq 'c' &&
      (!$cgi->param('time1') || !$cgi->param('time2')))
    ||
    ($cgi->param('time1') &&
     !($cgi->param('time1') =~ 
         /^[0-9]{4}\-[0-9]{2}\-[0-9]{2} [0-9]{2}\:[0-9]{2}$/))
    ||
    ($cgi->param('time2') &&
     !($cgi->param('time2') =~
         /^[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}$/))
    ||
    ($cgi->param('crrd') && !$cgi->param('cds'))
    ||
    ($cgi->param('cds') && !$cgi->param('crrd'))
    ||
    ($cgi->param('crrd') &&
     ($trrd = $cgi->param('crrd')) &&
     !(grep /^$trrd$/, @rrds))
    ||
    ($cgi->param('cds') && 
     !($cgi->param('cds') =~ /^[0-1]+$/))) {
  print $cgi->header();
  print "Bad parameter value\n";
  exit 0;
}


# ---------------- #
# Cookies & Header #
# ---------------- #

# Summary View cookie
if ($cgi->param('summary_view') && ($cgi->param('summary_view') =~ /^[hdbwmy]$/)) {
  $cookie = $cgi->cookie(
    -name=>'summary_view',
    -value=>$cgi->param('summary_view'),
    -expires=>gmtime(time()+365*24*3600)." GMT",
    -path=>'/',
  );
}

# RRD option cookie (which DS shall we display)
elsif ($cgi->param('crrd')) {
  my $temp = $cgi->cookie($cgi->param('crrd'));
  unless (($temp && $temp eq $cgi->param('cds')) ||
          (!$temp && ($cgi->param('cds') =~ /^1+$/)) ||
          ($cgi->param('cds') =~ /^0+$/)) {
    my $expires;
    if ($cgi->param('cds') =~ /^1+$/) {
      $expires = '-1d';
    } else {
      $expires = gmtime(time()+365*24*3600)." GMT";
    }
    $cookie = $cgi->cookie(
      -name=>''.$cgi->param('crrd'),
      -value=>$cgi->param('cds'),
      -expires=>$expires,
      -path=>'/',
    );
  }
}

# Header
if ($cookie) {
  print $cgi->header(-cookie=>$cookie);
} else {
  print $cgi->header();
}


# ------------------------ #
# Clean of old image files #
# ------------------------ #

my $now = time();
opendir my $dir, $GRA_DIR;
my @files = grep (/\.png/, readdir $dir);
closedir $dir;
foreach my $file (@files) {
  my @stats = stat($GRA_DIR.$file);
  if ($now - $stats[9] > $MAX_IMG_AGE) {
    unlink $GRA_DIR.$file;
  }
}


# ----------------------------------------- #
# rrd is passed in parameter to the page... #
# ----------------------------------------- #
if (defined $cgi->param('rrd') && $cgi->param('rrd') =~ /[a-zA-Z0-9]/) {

  my $rrd_id = $cgi->param('rrd');
  $rrd_id =~ s/^[0-9]+_//;

  # ------------ #
  # Zoom mode on #
  # ------------ #
  if ($cgi->param('type')) {

    # ----------- #
    # Normal mode #
    # ----------- #
    if ($cgi->param('type') ne 'c') {
      &createRRDGraph($cgi->param('rrd'),$cgi->param('type'));
      $title = $GRAPH_TITLES{$rrd_id};
      $back_link = 'index.pl?rrd='.$cgi->param('rrd');
      $ctime1 = strftime "%Y-%m-%d %H:%M", localtime($elgraphs[0]{'time1'});
      $ctime2 = strftime "%Y-%m-%d %H:%M", localtime($elgraphs[0]{'time2'});
    }

    # ---------------- #
    # Custom view mode #
    # ---------------- #
    else {
      &createRRDGraph($cgi->param('rrd'),'c',$cgi->param('time1'),$cgi->param('time2'));
      $title = $GRAPH_TITLES{$rrd_id};
      $back_link = 'index.pl?rrd='.$cgi->param('rrd');
      $custom_view = 1;
      $display_custom_view = 1;
      $ctime1 = $cgi->param('time1');
      $ctime2 = $cgi->param('time2');
    }
    $display_custom_view = 1;
  }

  # ------------- #
  # Zoom mode off #
  # ------------- #
  else {

    &createRRDGraph($cgi->param('rrd'),'h');
    &createRRDGraph($cgi->param('rrd'),'d');
    &createRRDGraph($cgi->param('rrd'),'b');
    &createRRDGraph($cgi->param('rrd'),'w');
    &createRRDGraph($cgi->param('rrd'),'m');
    &createRRDGraph($cgi->param('rrd'),'y');

    $title = $GRAPH_TITLES{$rrd_id};
    $back_link = 'index.pl';
    $display_custom_view = 1;

    $ctime1 = DateTime->from_epoch(epoch => time() - 172800, time_zone => "local");
    $ctime2 = DateTime->from_epoch(epoch => time() - 86400, time_zone => "local");
    $ctime1 = $ctime1->strftime('%F %H:%M');
    $ctime2 = $ctime2->strftime('%F %H:%M'); 
  }
}


# ----------------------------------------- #
# rrd isn't passed in parameter to the page #
# ----------------------------------------- #
else {

  my $user_view = $cgi->param('summary_view');
  $user_view = $cgi->cookie('summary_view') unless $user_view =~ /^[hdbwmy]$/;
  $user_view = $DEFAULT_VIEW unless $user_view =~ /^[hdbwmy]$/; 
  $summary = $user_view;
  foreach my $rrd (@rrds) {
    createRRDGraph($rrd, $user_view);
  }
}


# ------------------- #
# Displaying template #
# ------------------- #
my $template = HTML::Template::Expr->new(
  filename => $TEM_DIR.'index.html',
  die_on_bad_params => 0,
  global_vars => '1',
  loop_context_vars => '1'
);
$title =~ s/\{#server#\}/$SERVER/g;
$template->param(template_path => $TEM_DIR);
$template->param(server => $SERVER);
$template->param(title => $title);
$template->param(back_link => $back_link);
$template->param(summary => $summary);
$template->param(display_custom_view => $display_custom_view);
$template->param(custom_view => $custom_view);
$template->param(ctime1 => $ctime1);
$template->param(ctime2 => $ctime2);
$template->param(elgraphs => \@elgraphs);
$template->param(rrd => $cgi->param('rrd'));
$template->param(zoom_mode => defined $cgi->param('rrd') &&
  $cgi->param('rrd') =~ /[a-zA-Z0-9]/ && $cgi->param('type'));
unless ($summary) {
  $template->param(
    rrds => &getRRDItems($cgi->param('rrd')));
  $template->param(
    curl => '&type='.$cgi->param('type').
            '&time1='.$cgi->param('time1').
            '&time2='.$cgi->param('time2'));
}
print($template->output);
foreach my $trace (@trace) {print $trace;}
